### ISSUE TITLE
Expand DynamicDependencyAttribute handling with interfaces

### ISSUE THREAD

**Issue by marek-safar** [mono/linker#1636]
We have the following cases today which have to use linker descriptors and dependencies are always kept even though they are rarely used and should be trimmed to reduce the size.

All cases are a mix of managed and native code which need to cooperate but linker has no visibility to what native code requires/depends on.

## ICustomMarshaler case

We need to keep `ICustomMarshaler` methods and their implementations for any type which exists in the final output (we could go beyond that but that's probably stretch). Then if any of the interface methods are kept we need to also keep managed logic in `Marshal.GetCustomMarshalerInstance` method.

Proposed extensions

```c#
[DynamicDependencyAttribute("MarshalNativeToManaged")]
[DynamicDependencyAttribute("MarshalManagedToNative")]
[DynamicDependencyAttribute("CleanUpNativeData")]
[DynamicDependencyAttribute("CleanUpManagedData")]
[DynamicDependencyAttribute("GetNativeDataSize")]
public interface ICustomMarshaler
{
	// Keep the dependency if it's never used in managed code
	[DynamicDependencyAttribute("Marshal.GetCustomMarshalerInstance")]
    object MarshalNativeToManaged(IntPtr pNativeData);
    IntPtr MarshalManagedToNative(object ManagedObj);
    void CleanUpNativeData(IntPtr pNativeData);
    void CleanUpManagedData(object ManagedObj);
    int GetNativeDataSize();
}
```

## Implicit Array interfaces

For MonoVM we have implementations of implicit array method done via magic mapping between vtable and managed method implementation. This goes a bit against the case above because ideally, we would like to keep the methods only if they are used.

Proposed extension

```c#
public interface IReadOnlyCollection<out T> : IEnumerable<T>
{
    [DynamicDependency("InternalArray__IReadOnlyCollection_get_Count", typeof(Array))]
    int Count { get; }
}
```

This one is trickier as ideally the dependency should be kept only when Array is kept but in this particular case that's very unlikely.

@vitek-karas

**Comment by vitek-karas** [comment]
Just to clarify:

* In the custom marshaler case the proposed changes are not enough for the scenario to work, right? Since the type which implements the interface also needs its `.ctor`.

* In the array case, I assume you meant that optimizing to only have the dependency when `Array` is also included is not that important since every app will keep `Array` anyway, right?

**Comment by vitek-karas** [comment]
Allowing `DynamicDependency` on type makes sense to me. The meaning would be:
"If the attributed type is kept in the linker output, then the target of the `DynamicDependency` will also be kept.`

*Note: It says nothing about the type being instantiated, in implementation terms, just the fact that something called `MarkType(TypeReference)` will be enough to trigger the `DynamicDependencies` processing to happen.*

As for the proposed, it would mean these changes to the linker (and runtime):
- [ ] Allow `DynamicDependencyAttribute` on `AttributeTarget.Class`, `AttributeTarget.Struct` and `AttributeTarget.Interface` - I would do it on all "types", it's not worth the trouble if we only focus on interface and later on find we need it elsewhere. From linker's point of view it makes very little difference.
- [ ] Implement support for `DynamicDependencyAttribute` on types in the linker
  - The initial implementation could be made without changes to the `DynamicDependencyLookupStep` - it's not worth doing that part since it will go away once https://github.com/mono/linker/issues/1164 is done.
- [ ] Verify that `DynamicDependencyAttribute` on an interface method actually works
  - [ ] If not fix it - the semantics should be the same as for any other method - if the method definition (there's no body in this case) is kept, then the target of `DynamicDependencyAttribute` should be kept as well.
  - [ ] Also verify that it works on default interface methods
- [ ] (Not really part of this, but still worth doing with it) Verify that creating dependency loops via `DynamicDependencyAttribute` doesn't break the linker in bad ways - linker should keep the entire loop if any of it's members are kept. This should include self-referencing dependencies.

As mentioned above, this doesn't solve the potential need for a dependency like:
"If type A is kept and it implements interface IB, then the `.ctor` of A needs to be kept as well". If it's confirmed that such dependency is also needed, let's create a separate issue for it.

**Comment by marek-safar** [comment]
> In the custom marshaler case the proposed changes are not enough for the scenario to work, right? Since the type which implements the interface also needs its .ctor.

Yes, that's my understanding. This might change in the future if we ever add marshalling code generators but right now the use case is like
https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.icustommarshaler?view=net-5.0#using-a-custom-marshaler

We could fix the ctor case by annotating https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshalasattribute.marshaltyperef?view=net-5.0 type to require public ctor but not sure if we can do the same for https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshalasattribute.marshaltype?view=net-5.0.

> to only have the dependency when Array is also included is not that important since every app will keep Array anyway, right

Correct but that's the Array case if we apply the same pattern elsewhere it might matter.

**Comment by sbomer** [comment]
If I follow, the proposal is to put `DynamicDependency` on `ICustomMarshaler` to keep the interface methods. Will we also use `DynamicDependency` on types that implement `ICustomMarshaler` to keep the interface implementation? Or do we need a mechanism to automatically keep the `ICustomMarshaler` implementation also for unannotated types?

**Comment by marek-safar** [comment]
> Will we also use DynamicDependency on types that implement ICustomMarshaler to keep the interface implementation?

Why should that be needed? Marking an interface method should already keep method implementations if the type can be constructed.

**Comment by sbomer** [comment]
My understanding is that the interface methods will be marked by `DynamicDependency` on the marked interface - is that right? What will mark the interface?

The interface removal optimization could remove interfaces even for constructed types if the interface isn't referenced from used code. If there is always going to be a used reference to the interface (like a call through the interface method) it should be fine.

For example:
```csharp
// IFoo gets removed
interface IFoo {
    void M();
}

class C : IFoo {
    public void M() {
        Console.WriteLine("Hey");
    }
}

class Program {
    static void Main(string[] args) {
        new C().M();
        // Helper(); // uncommenting this will keep IFoo (but not its interface methods)
    }

    static void Helper() {
        var t = typeof(IFoo);
    }
}
```

**Comment by marek-safar** [comment]
Please see my comment about ctor marking in https://github.com/mono/linker/issues/1636#issuecomment-731234552

**Comment by MichalStrehovsky** [comment]
I think there's some confusion about how ICustomMarshaler works - there's some discussion of that in #1281 and the associated pull request and I think it gives answers to all questions (we don't need to mark default constructors, annotating MarshalAs attribute is not possible because it's not a real attribute, and we're going to keep the interface because it's the return type of the GetInstance method that we keep due to linker special casing).

The ICustomMarshaler case could also be solved by simply extending the existing special casing within the linker to also keep all the interface methods. Hardcoding fully documented details into linker is not too bad. The class library "knows" about these details as much as linker (this interface is not a CoreLib concern either and dumping knowledge into CoreLib so that linker doesn't delete stuff the runtime needs feels like one extra place that needs to become aware of this that maybe doesn't have to)

**Comment by vitek-karas** [comment]
My understanding was that @marek-safar knows about other places in Xamarin where similar problems occur. Marshalling was just and example I think. But I'll let him speck to that...

**Comment by marek-safar** [comment]
I'll create separate issues for other scenarios to not complicate this further but we need to solve the problems described above. If you agree on that the recommended solution is to hardcode the knowledge to linker I'm fine with that.

**Comment by vitek-karas** [comment]
This is a public contract, it's basically similar to linker having knowledge about reflection to some degree. So I'm OK with hardcoding the rules into linker for this.
